core/src/main/java/org/eclipse/microprofile/reactive/streams/core/CompletionRunnerImpl.java

```diff
@@ -0,0 +1,50 @@
+/*******************************************************************************
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+package org.eclipse.microprofile.reactive.streams.core;
+
+import org.eclipse.microprofile.reactive.streams.CompletionRunner;
+import org.eclipse.microprofile.reactive.streams.spi.Graph;
+import org.eclipse.microprofile.reactive.streams.spi.ReactiveStreamsEngine;
+import org.eclipse.microprofile.reactive.streams.spi.Stage;
+
+import java.util.Objects;
+import java.util.concurrent.CompletionStage;
+
+final class CompletionRunnerImpl<T> extends ReactiveStreamsGraphBuilder implements CompletionRunner<T> {
```

Comment:
I wondered why some impl classes like this one are final?

Reply:
They should all be final. In theory, there could be performance benefits to making things final, though I believe in practice this actually isn't true in Hotspot at least, but that could change for future versions of the JVM, and I don't know whether J9 is able to infer effective finality like Hotspot is able to.

But really, it's a matter of code hygiene, marking a class as final indicates to whoever is reading the code that there are no subclasses of this class, and so they can be sure that when a method is invoked on the class, the implementation they see in front of them is what is executed, they don't have to take into account that it could be overridden by some subclass that they haven't seen yet. This reduces the number of variables that a developer needs to keep in their head as they read and understand the code, making the code easier to reason about. It also shows intent, this class has not been designed to be extended, and so anyone making changes in future should be very wary if they change it to not be final and try to extend it.